### ISSUE TITLE
[Godot] Support for 3D system

### ISSUE THREAD

**Issue by julhe** [EsotericSoftware/spine-runtimes#2135]
It would be nice if the Godot integration supports the 3D system as well. The game we're making is 2D, but Godots 2D system is limited, forcing us to switch to 3D.

By the way, I'm currently maintaining my own branch for this: https://github.com/julhe/spine-runtimes which "works", but a official version would be really cool at some point.

**EDIT: Please don't post stuff like "+1". Just upvote, please for the sake of my inbox**

**Comment by badlogic** [comment]
I considered this for the V1, but as you found in your fork, it's not exactly trivial to get feature parity with the 2D version. It's definitely planned, but not in the short term I'm afraid.

**Comment by rinvars** [comment]
+1 for 3D support

**Comment by NotBadCode** [comment]
+1

**Comment by shiouhuamien** [comment]
+1 Very necessary
I also want to make 2.5D games

**Comment by sun-studios** [comment]
+1 2D/3D hybrid games are my jam

**Comment by jpeinelt** [comment]
same here, 2.5D games have this special look and spine is awesome, so we need to combine both :)

**Comment by BunkWire2X8** [comment]
Yeah, I've been looking forever for 2D puppet animation in 3D support for my Godot projects. The old one I used was DragonBones for Unity, but I can't stick with that one now due to certain recent issues regarding that game engine.

**Comment by josuechopite** [comment]
Is this feature still planned? When do you plan to add it? Thanks.

**Comment by rafaskb** [comment]
I would also like to signal our needs for this feature. Our studio plans are to migrate from libGDX to Godot, but we're working with 3D scenes to create 2.5D games, and having Spine to work with 3D is a must. Really looking forward to this feature, hopefully it gets a priority bump 😊

**Comment by badlogic** [comment]
Folks, everytime you post a "+1" anyone subscribed to the repository gets an email. Could you please instead simply use the "upvote" emoji on the original comment in this issue?

I know this is of importance, but we are a small team and can only do so much given our resources. We couldn't anticipate the influx of Unity users to Godot given recent news, and we are trying our best to get the missing features in spine-godot going for spine-unity users. But we can't simply put all the work for all the other runtimes on the back burner immediately. Resource planning and such, you know :)

**Comment by brogan89** [comment]
I did manage to get something working with SubViewport. I'm guessing people want this for an HD-2D game or any kind of 2.5D game.

I achieved it by adding a `MeshInstance3D` and setting mesh to a quad. Add a material override, and set Transparency to `Alpha Scissor`. This is so shadows can be cast. 

Add a `SubViewport` as a child node. Set `Transparent BG` to true. 

Add a `SpineSprite` node as a child to `SubViewport` node. Set the position to the middle of the SubViewport texture (256, 512) if the SubViewport size is 512x512. 

Add a script to the `MeshInstance3D` node 

```gdscript
@tool
extends MeshInstance3D

func _ready():
	if material_override != null:
		material_override.albedo_texture = $SubViewport.get_texture()
```

Done. Works fine. Not sure if it's very scalable as I'm sure there is overhead to this approach, but it'll get me by as I prototype

![image](https://github.com/EsotericSoftware/spine-runtimes/assets/8637789/81a33611-6269-4f4f-846d-b851da387a98)

![image](https://github.com/EsotericSoftware/spine-runtimes/assets/8637789/054fe875-603f-4039-bf9a-46e3e309d606)

Hope that helps anyone coming across this.

**Comment by 80dots** [comment]
+1

**Comment by yangqingming** [comment]
+1

**Comment by yangqingming** [comment]
+1

**Comment by emarino135** [comment]
+1

**Comment by emarino135** [comment]
Any update on if this will be added anytime soon?

**Comment by luanhadev** [comment]
Is there any update? I’m still waiting for this feature.

**Comment by brogan89** [comment]
> Is there any update? I’m still waiting for this feature.

Why are you waiting for the feature? 
- If you're making a game for production that requires this feature you shouldn't use Godot. Use another engine.
- If you're making a game for hobby and/or wanting to learn, then there is no rush, you can wait.

Plus you can already get this feature with a workaround, I'm already making a game for production without needing the feature. [The workaround I've posted](https://github.com/EsotericSoftware/spine-runtimes/issues/2135#issuecomment-1736450408)

This is Open Source, they will get it when they're ready, or until someone submits a PR adding the feature. 

So if you want the feature, you gotta get in there and get your hands dirty and add it yourself, or patiently wait. 

Everyone posting "Any update, is it ready yet?" etc. Is not helping, Its clearly not done yet as the issue is still Open.

EDIT
The OP even has his own fork with the fix, have you tried that?
I didn't notice it until now, I'll definitely try that out soon. lol
@julhe you should submit PR for this :)

**Comment by HKunogi** [comment]
+1

**Comment by N-Saint** [comment]
+1

**Comment by mbbdogg** [comment]
yeah sorry bro im plus 1ing this. +1

**Comment by 500nights** [comment]
This is a fundamental feature that I feel like should have upmost priority especially after how much time has passed, we’ve been waiting on this for quite some time and while there are alternatives… they aren’t great solutions. I’m actually a bit surprised how long this has taken the as the spine team is incredibly talented. I mean their software alone is incredible and beats out any competition that even attempts to come close, but even more so their technical support and support for run times are absolutely top notch, the best I’ve seen from any software development team. Hopefully we get a proper implementation soon as I know a lot of us are looking forward to those 😋. Looking forward to waving goodbye to the days of janky viewport and mesh3d shenanigans!

**Comment by cmr624** [comment]
Hi all, we just implemented a [3D game with SpineSprites](https://crabonfireco.itch.io/trait-tycoon) rendering animated characters.

You can see the [structure of the basic unit](https://github.com/violent-coyote/gmtk-24/blob/main/scenes/GameScene/Game/unit_base.tscn) we used to create this effect, pulled from solutions posted earlier in this thread. 

We also built a version of Godot with the [spine runtimes](https://github.com/violent-coyote/spine-runtimes) that works with [Godot 4.3-stable](https://github.com/violent-coyote/spine-runtimes/actions/runs/10416123447), as the spine official runtimes only go up to 4.2.2-stable.

It's a very feasible option for small prototypes or projects, but we noticed the spine sprites don't render in every browser on web.

**Comment by giotiskl** [comment]
Got this working with the workarounds in the thread but an official proper solution would be highly appreciated 👍🏻

**Comment by bitsalive-ron** [comment]
+1

**Comment by bitsalive-ron** [comment]
I would like to see that feature supported officially as well. 
Gonna try the workaround for now. Thanks for posting it!
Are people using this in a finished project already?

**Comment by Skroopa** [comment]
The only reason why I use UE5 instead of Godot, is only because of the missing 3D runtimes for Spine animations. Any movement on this?

**Comment by brogan89** [comment]
> The only reason why I use UE5 instead of Godot, is only because of the missing 3D runtimes for Spine animations. Any movement on this?

Just use the workaround for now. https://github.com/EsotericSoftware/spine-runtimes/issues/2135#issuecomment-1736450408
They will be working on it soon. https://github.com/EsotericSoftware/spine-runtimes/issues/2654#issuecomment-2410471882

**Comment by Skroopa** [comment]
> > The only reason why I use UE5 instead of Godot, is only because of the missing 3D runtimes for Spine animations. Any movement on this?
> 
> Just use the workaround for now. [#2135 (comment)](https://github.com/EsotericSoftware/spine-runtimes/issues/2135#issuecomment-1736450408) They will be working on it soon. [#2654 (comment)](https://github.com/EsotericSoftware/spine-runtimes/issues/2654#issuecomment-2410471882)

yes, I read all this already. I switched to frame-by-frame animation for now

**Comment by Minq-git** [comment]
> I did manage to get something working with SubViewport. I'm guessing people want this for an HD-2D game or any kind of 2.5D game.
> 
> I achieved it by adding a `MeshInstance3D` and setting mesh to a quad. Add a material override, and set Transparency to `Alpha Scissor`. This is so shadows can be cast.
> 
> Add a `SubViewport` as a child node. Set `Transparent BG` to true.
> 
> Add a `SpineSprite` node as a child to `SubViewport` node. Set the position to the middle of the SubViewport texture (256, 512) if the SubViewport size is 512x512.
> 
> Add a script to the `MeshInstance3D` node
> 
> ```gdscript
> @tool
> extends MeshInstance3D
> 
> func _ready():
> 	if material_override != null:
> 		material_override.albedo_texture = $SubViewport.get_texture()
> ```

Thank you so much @brogan89, this took a bit of learning but it worked great! My next step is to try mapping the `MeshInstance3D` to a `CharacterBody3D`.

Also, for anyone who may be using C# versions of the spine-godot, I have ported Brogan's original script to C# and added some error handling. Hope this also helps someone in the future!
```csharp
using Godot;

[Tool]
public partial class SpineSprite3D : MeshInstance3D
{
    public override void _Ready()
    {
        base._Ready();

        if (MaterialOverride == null)
        {
            GD.PrintErr("MaterialOverride is null. Ensure it is set in the inspector or by code.");
            return;
        }

        var subViewport = GetNode<SubViewport>("SubViewport");
        if (subViewport == null)
        {
            GD.PrintErr("SubViewport node not found. Check the node path.");
            return;
        }

        MaterialOverride.Set("albedo_texture", subViewport.GetTexture());
    }
}
```

**Comment by jeremeyzela** [comment]
Just here to add my +1 to this. I have a project with a bunch of spine content that I'm hoping to convert from Unreal to godot.

**Comment by rafaskb** [comment]
Any news on this request? It's been almost 3 years 😭

**Comment by yulrun** [comment]
Following up on this as well.

Currently looking into switching from full 3D to a 2.5D perspective to lighten workload as I'm a solo dev. I'm deeply embedded into Godot so won't look at changing engine but also have owned Spine2D pro for like 5 years so would like to make use of it.

I'll look at doing work around if no official commit is placed, but curious on the progress of this.

Cheers

**Comment by BroofTerr** [comment]
Still no news? 😭

**Comment by emarino135** [comment]
No

**Comment by luanhadev** [comment]
I've been eagerly waiting for the 3Dsprite version and checking this topic every month. I just saw a Moho video about exporting to Blender and wanted to share it here—maybe it'll give the devs some ideas? (Specifically, treating the 3Dsprite as a multi-mesh object and rigging it).
https://www.youtube.com/watch?v=mHkONlwP8co

**Comment by yulrun** [comment]
Now that it had a Standard GD extension runtime hopefully we can see this tackled soon!

**Comment by shoyguer** [comment]
Currently working on a PR to solve this.
Only adding SpineSprite3D, though. In the future, I may add other nodes too.

**Comment by yulrun** [comment]
> Currently working on a PR to solve this. Only adding SpineSprite3D, though. In the future, I may add other nodes too.

I feel having SpineSprite3D should solve majority of everything people would require. So thank you for this.

**Comment by AlorxGammer** [comment]
Hey! I have been following this issue for a while, and I ended up working on my own Godot 3D implementation for Spine.

I finally made the repository public today:

https://github.com/AlorxGammer/spine-runtimes-Godot-3D

It is an unofficial Godot 4.3+ GDExtension that lets Spine content live directly in a 3D scene, instead of using the usual SubViewport-to-quad workaround.

Right now it has separate branches for Spine 3.8, 4.0, 4.1, 4.2, and 4.3, plus GDScript/C# example projects and build scripts.

It is still actively being developed and there are things I want to improve, but the core functionality is already working and usable. I mainly wanted to share it here in case it helps other people who have been waiting for Godot 3D support.

Feedback or bug reports are welcome in the repository Issues.

**Comment by badlogic** [comment]
This is now in the 4.3 branch. See the [SpineSprite3D docs](https://en.esotericsoftware.com/spine-godot#SpineSprite3D-Node).
